lib/sequel/extensions/connection_metrics.rb

```diff
@@ -0,0 +1,100 @@
+      case pool.pool_type
+      when :single, :sharded_single, :timed_queue, :sharded_threaded, :sharded_timed_queue
+        raise Error.new('cannot load connection_metrics extension if using a connection pool type different than :threaded')
+      end
+
+      pool.instance_exec do
+        sync do
+          @prometheus_updater = CloudController::DependencyLocator.instance.prometheus_updater
+          @connection_info = {}
+        end
+      end
+    end
+
+    private
+
+    def acquire(thread)
+      begin
+        if (conn = super)
+          acquired_at = Time.now
+          @prometheus_updater.increment_gauge_metric(:cc_acquired_db_connections_total, labels: { process_type: })
+          @connection_info[thread] ||= {}
+          @connection_info[thread][:acquired_at] = acquired_at
+        end
+      rescue Sequel::PoolTimeout
+        @prometheus_updater.increment_gauge_metric(:cc_db_connection_pool_timeouts_total, labels: { process_type: })
+        raise
+      ensure
+        # acquire calls assign_connection, where the thread possibly has to wait for a free connection.
+        # For both cases, when the thread acquires a connection in time, or when it runs into a PoolTimeout,
+        # we emmit the time the thread waited for a connection.
+        if @connection_info[thread] && @connection_info[thread].key?(:waiting_since)
+          @prometheus_updater.update_histogram_metric :cc_db_connection_wait_duration_seconds, (Time.now - @connection_info[thread][:waiting_since]).seconds
+          @connection_info[thread].delete(:waiting_since)
+        end
+      end
+      conn
+    end
+
+    def assign_connection(thread)
+      conn = super
+
+      # if assign_connection does not return a connection, the pool is exhausted and the thread has to wait
+      unless conn
+        waiting_since = Time.now
+        @connection_info[thread] = { waiting_since: } unless @connection_info[thread] && @connection_info[thread].key?(:waiting_since)
+      end
+      conn
+    end
+
+    def make_new(server)
+      conn = super
+      @prometheus_updater.update_gauge_metric(:cc_open_db_connections_total, size, labels: { process_type: })
+      conn
+    end
+
+    def disconnect_connection(conn)
+      super
+      @prometheus_updater.update_gauge_metric(:cc_open_db_connections_total, size, labels: { process_type: })
+    end
+
+    def release(thread)
+      super
+
+      # acquired_at should be always set, but as a safeguard we check that it is present before accessi
+      if @connection_info[thread] && @connection_info[thread].key?(:acquired_at)
+        @prometheus_updater.update_histogram_metric :cc_db_connection_hold_duration_seconds, (Time.now - @connection_info[thread][:acquired_at]).seconds
+      end
+
+      @prometheus_updater.decrement_gauge_metric(:cc_acquired_db_connections_total, labels: { process_type: })
+      @connection_info.delete(thread)
+    end
+
+    def process_type
+      ENV['IS_MAIN'] == 'true' ? 'main' : 'worker'
```

Comment:
**thought** Could `worker` here be confused with CC worker as opposed to puma worker?

Reply:
@Samze, would something like 'web_worker' help to avoid potential confusion?

Reply:
I renamed it to `puma_worker`. Currently the CC workers do not emit those metrics, but if they did in the future it might be confusing indeed.